text/0000-generalized-type-ascription.md

```diff
@@ -0,0 +1,2441 @@
+writing:
+
+```rust
+trait Foo {
+    fn bar(MyType) -> ... { ... }
+}
+```
+
+will cause the compiler to emit a warn-by-default lint in Rust 2015 and
+that it be a hard error in Rust 2018. This resolves the unresolved question
+in [RFC 1685].
+
+## Linting ascription of named `struct` literals and patterns
+
+Consider a struct:
+
+```rust
+struct Foo<T> {
+    bar: T
+}
+```
+
+When it comes to type ascribing the field `bar` in a struct literal expression
+such as `Foo { bar: x : Type }` or in particular when you type ascribe
+`bar` in a pattern: `Foo { bar: x : Type }` it is not always very clear from
+this way of writing what is what.
+
+We propose therefore that the compiler should provide a warn-by-default lint
+that suggests that you should wrap the ascription in parenthesis like so:
+
+```rust
+let x = Foo { bar: (x : Type) }
+
+let Foo { bar: (x: Type) } = ...;
+```
+
+This lint only applies when after giving fresh names for all identifiers inside
+`x` and `Type`, their token streams match (α-equivalence).
+For example, this means that if you write `let Foo { bar: x : u32 }`
+or `let Foo { bar: &x : &X }` the compiler will emit a warning.
+However, if you write `let Foo { bar: x : Vec<u8> }` it will not.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Grammar
+
+Assuming the following tokens and the terminal strings they lex:
+
+```rust
+LOOP : 'loop' ;
+UNSAFE : 'unsafe' ;
+TRY : 'try' ;
+ASYNC : 'async' ;
+LET : 'let' ;
+SELF : 'self' ;
+DOTDOTDOT : '...' ;
+```
+
+The following alternatives are modified in the expression grammar:
+
+```rust
+ascribe: ':' ty_sum ;
+
+expr
+: ...
+| expr ascribe // This is specified in RFC 803 but it is included for completeness.
+| LOOP ascribe? block // This replaces the existing production for loop { .. }.
+| UNSAFE ascribe? block
+| TRY ascribe? block
+| ASYNC ascribe? block
+;
+```
+
+Here, the precedence of `:` in the alternative `expr ascribe` is the same as
+the operator `as`. However, when the parser encounters type ascription of an
+expression immediately followed by a field projection or a method call,
+then the parser shall interpret the projection and the call as being performed
+on the ascribed expression. Thus, if a user writes `expr : type . field`
+the parser associates this as `(expr : type) . field`. Similarly, if a user
+writes `expr : type . method(..)` the parser associates this as
+`(expr : type) . method(..)`. An implementation of this wrt. method calls
+exists in [rust-lang/rust#33380](https://github.com/rust-lang/rust/pull/33380).
+
+To the pattern grammar, the following alternative is added:
+
+```rust
+pat
+: ...
+| pat ascribe
+;
+```
+
+The operator `:` binds more tightly than `ref` and `ref mut` but binds less
+tightly than `&` and `&mut` in pattern contexts. This is required because
+currently, the following compiles:
+
+```rust
+#[derive(Copy, Clone)]
+struct X {}
+let a = X {};
+
+let ref b: X = a; // Note the type!
```

Comment:
This is such an annoying detail.
This is required for backward compatibility, but otherwise goes against all pattern rules and intuition.
`ref` can't even have an arbitrary pattern on its right side, so `IDENT: TYPE` in `ref IDENT: TYPE` is some kind of special syntactic construction rather than a pattern.

Reply:
(Disclaimer: I don't know how `let ref` interacts with more complex patterns than `b` or `b : X`.)

You mention below, that it's interpreted as `ref (x : T)`. I disagree. I you regard the `ref`-operator as a **dereferencing** operator (which I'm afraid is counterintuitive because of its name), then I'd argue it's interpreted `(ref x) : T`. `x` is a reference and if you dereference it (i.e. take `ref x`), then you get a `T`. Therefore we have `(ref x) : T`. 

(The mentioned fact that `x` is a reference is infered from that the `ref`-operator is applied to it and references can be dereferenced.)

If my interpretation happens to be true, then the whole interaction between type ascription and the `ref`-operator should of course be designed around that interpretation.

### Alternative

As of today, is the `ref`-operator still needed at all? If it could be made deprecated, then there would be no need to support type ascription for expressions using it.

### Anecdotal

Regarding `ref` as a dereferencing-operator, it would of course be more consistent if this syntax used the dereferencing operator which is also used in expressions: `*`. Would that collide with the current pointer syntax? Where can I read more about these decisions?

Reply:
@haslersn 

> You mention below, that it's interpreted as `ref (x : T)`. I disagree. I you regard the `ref`-operator as a **dereferencing** operator (which I'm afraid is counterintuitive because of its name), then I'd argue it's interpreted `(ref x) : T`. `x` is a reference and if you dereference it (i.e. take `ref x`), then you get a `T`. Therefore we have `(ref x) : T`.

Interesting. This could be better because clearly the type of `x` is `&T` here, not `T`; that should also avoid the special syntactic construct. The idea would be for `ref x` to have not the type of `x`, but the type `x` refers to. This also works well with:

```rust
fn _foo() {
    struct X;
    let a = (X, X);
    let (ref _b, ref _c): (X, X) = a;
}
```
which type checks today...

@nikomatsakis ^-- does this work out with subtyping trickyness?

> ### Alternative
> 
> As of today, is the `ref`-operator still needed at all? If it could be made deprecated, then there would be no need to support type ascription for expressions using it.

Sure, it's used today by people who wish to be more explicit wrt. match bindings and who don't wish to use default match bindings. We have no plans to deprecate `ref mut? x`.

Reply:
Occurences of `ref mut x` possibly need to be changed to `mut ref x` because it's the `ref x` (i.e. the value `x` refers to) that's `mut`.

Reply:
@haslersn I don't believe `mut ref x` is a valid pattern today; `ref mut x` is a special construction, it's not `ref (mut x)`, but rather `(ref mut) x` so there should be no problem... An example:
```rust
fn _foo() {
    struct X;
    let mut a = (X, X);
    let (ref mut _b, ref mut _c): (X, X) = a;
}
```